src/line_cursor.ts

```diff
@@ -673,22 +709,16 @@ export class LineCursor extends Marker {
   /**
    * Event listener that syncs the cursor location to the selected
    * block on SELECTED events.
+   *
+   * @param event The `Selected` event.
    */
   private selectListener(event: Blockly.Events.Abstract) {
     if (event.type !== Blockly.Events.SELECTED) return;
     const selectedEvent = event as Blockly.Events.Selected;
     if (selectedEvent.workspaceId !== this.workspace.id) return;
-    if (selectedEvent.newElementId) {
-      const block = this.workspace.getBlockById(selectedEvent.newElementId);
-      if (block) {
-        const node = ASTNode.createBlockNode(block);
-        if (node) {
-          this.setCurNode(node, true);
-        }
-      }
-    } else {
-      this.setCurNode(null as never, true);
-    }
+    // This runs too late so the logic to update the selection is in
+    // `getCurNode`.
```

Comment:
The context here is missing. It makes sense when reading the PR, but this comment by itself post-merge will be hard to understand without blame. Can more context perhaps be added?

Separately, might it be possible to factor out the bits of `getCurNode` which actually update the state into a common helper called both there and here? That might be the cleanest way to indicate what needs to happen in both locations.

Reply:
That really helped. 1233f578fd94a0941a8e028a0417d7e83a3eda3f has separate methods for the node -> selection and selection -> node updates and uses the selection -> node one in the listener and `getCurNode`.

Reply:
Ultimately I hope https://github.com/google/blockly/issues/8820 might make this less messy

Reply:
I find the latest changes much cleaner--thanks!